Allow specifying the primary group id of the container

krmayankk opened the pull request:
Allows specifying the Primary Group ID of the container, in addition to Primary User Id when running in Docker similar to how docker supports this.

krmayankk commented:
@pmorie @thockin @pineking @vpal @jheiss

krmayankk commented:
@kubernetes/sig-node-proposals

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,112 @@
+## Abstract
+
+
+As a Kubernetes User, i should be able to specify both user id and group id for the containers running 
```
s/i/I/

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,112 @@
+inside a pod on a per Container basis, similar to how docker allows that using docker run options -u, 
+--user="" Username or UID (format: <name|uid>[:<group|gid>]) format.
+
+PodSecurityContext allows Kubernetes users to specify RUnAsUser which can be overriden by RunAsUser
```
typo: RUnAsUser

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,112 @@
+--user="" Username or UID (format: <name|uid>[:<group|gid>]) format.
+
+PodSecurityContext allows Kubernetes users to specify RUnAsUser which can be overriden by RunAsUser
+in SecurtiyContext on a per Container basis. There is no equivalent field for specifying the primary
```
typo: SecurityContext

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,112 @@
+## Motivation
+
+Enterprise Kubernetes users want to run containers as non root. This means running containers with a 
+non zero user id and non zero group id. This gives Enterprises confidence that their customer code
```
Group zero essentially means nothing and has zero security implications.  I'm not saying that this isn't a valid use case, but you don't gain anything, to my knowledge, by running as GID 0.

pmorie replied:
I would buy that someone wants to be able to use GIDs as part of a security scheme, though, possibly derived from LDAP or another identity system.

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,112 @@
+## Use Cases
+
+Use case 1:
+As a cluster operator, i should be able to control both user id and primary group id of containers 
```
I think this is 'as a Kubernetes user', and what's the "so that I can _____________" part of this use-case?

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,112 @@
+launched using Kubernetes.
+
+Use case 2:
+As a cluster operator, i should be able to override user id and primary group id of a container as 
```
Isn't this the same use case as the first one?

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,112 @@
+ }
+ 
+
+type PoddSecurityContext struct {
```
typo: `PodSecurityContext`

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,112 @@
+     // +optional
+     RunAsUser *int64
+     // The GID to run the entrypoint of the container process.
+     // Defaults to user specified in image metadata if unspecified.
```
Defaults to the group?

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,112 @@
+- If no RunAsGroup is provided in the PodSecurityContext and SecurityContext, and none in the image,
+  the container will run with primary Group as root(0).
+
+## Not In Scope
```
How is this not in scope?  We need to add the policy control surface to govern features like this as we implement them.

smarterclayton replied:
Any additions to PodSecurityContext will changes to PodSecurityPolicy - both are part of the API and should evolve together.

erictune commented:
Can you at least compare to the alternative, which is pushing an /etc/groups file to all your nodes and using that for group membership?

krmayankk commented:
@pmorie @smarterclayton addressed comments. @erictune Are you suggesting controlling the gid through the file, while controlling uid through existing runtime mechanisms. It can be done, but not dynamically configurable and one /etc/groups file will affect all containers of a host. Do you still think we should discuss that option in the proposal ?

krmayankk commented:
@erictune forgot to mention that /etc/groups is only for secondary group ids. Here we are talking about controlling the primary group id the container runs as

vpal commented:
@krmayankk, @erictune  I think that the host file on the host OS does not affect UID and GID allocation in the container.
```
sudo docker run --rm -ti --user 1000 ubuntu:16.04 /bin/bash
I have no name!@80925fc54764:/$ id
uid=1000 gid=0(root) groups=0(root)
I have no name!@80925fc54764:/$ 
```
I have user and group 1000 in the group file on the host.
So what you would need to do is rebuild your images with the /etc/groups file, which is the thing we would like to avoid here. That is what the --user is for in docker.
If you have the chance to rebuild the image you can also solve this in other ways like providing UID and GID in the Dockerfile for example.
But if you just want to run something straight from a third party repository and want it to have a specific UID and GID overriding the default one you do not have the option right now. You can only specify the UID and supplemental groups.

smarterclayton commented:
I have no objections to this as stated.  For symmetry I see no reason why this should be prohibited.

smarterclayton commented:
Something seems to be wrong with lowercase 'i''s in the markdown being converted to capital I.

smarterclayton commented:
@kubernetes/sig-auth-feature-requests can you chime in if you see any additional implications.  I don't see any reason not to add this or enable it as long as the PSP work is done at the same time.

erictune commented:
Disregard my comment about /etc/groups.

On Tue, Aug 15, 2017 at 3:53 PM, Clayton Coleman <notifications@github.com>
wrote:

> @kubernetes/sig-auth-feature-requests
> <https://github.com/orgs/kubernetes/teams/sig-auth-feature-requests> can
> you chime in if you see any additional implications. I don't see any reason
> not to add this or enable it as long as the PSP work is done at the same
> time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/756#issuecomment-322611170>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHuudqchstCVUhYMmVuyd88UUTdIH9Rqks5sYiFYgaJpZM4OB0P7>
> .
>

smarterclayton commented:
I'd like at least another sig node reviewer to comment before I mark lgtm.

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,164 @@
+
+### Model
+
+Introduce a new API field in SecurityContext and PodSecurityContext called `RunAsGroup`
```
Missing period at the end of this sentence.  Can we mention explicitly what the behavior is if these fields are not defined?

krmayankk commented:
@pmorie there is a `Behavior` heading already at the bottom of this design proposal. Let me know if its missing something ?

pmorie commented:
@krmayankk The behaviors look well specified to me.  One thing I have overlooked until now is that I would like a summary of the code changes required to implement this.  For example, what changes are required for areas like:

- API validations
- Kubelet changes
- CRI interface changes
- Code changes associated with PSP
- Default PSP changes, if any.

krmayankk commented:
@pmorie i have added a summary of the areas that get affected and the folders by doing a grep analysis. PTAL. Also lot of tests will change or new will be added. As i get a deeper understanding of other areas that need changing, i will update the design.

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,191 @@
+into the system. Two of the relevant strategies are RunAsUser and SupplementalGroups. We introduce 
+a new strategy called RunAsGroup which will support the following options:
+- MustRunAs
+- MustRunAsNonRoot
```
I don't think we need `MustRunAsNonRoot`, since the root group has no connotations or elevated permissions associated with it.  WDYT @pweil-

krmayankk replied:
the root group has similar meaning as root user and the MustRunAsNonRoot would ensure group is always non zero the same way as MustRunAsNonRoot would ensure user is always non zero.

smarterclayton replied:
I'm fine with run as group and the changes here because there are cases where it matters to be not group zero and there's no dealt reason to not support it equally (less inconsistency).

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,191 @@
+- pkg/security/podsecuritypolicy/user
+- pkg/security/podsecuritypolicy/group
+- pkg/registry/extensions/podsecuritypolicy/storage
+- pkg/kubelet/rkt
```
Does rkt support setting the primary group?

krmayankk replied:
from the doc, it seems it does https://coreos.com/rkt/docs/latest/subcommands/run.html Look at the section `Overriding User/Group`

pmorie commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,191 @@
+- pkg/security/podsecuritypolicy/selinux
+- pkg/security/podsecuritypolicy/user
+- pkg/security/podsecuritypolicy/group
+- pkg/registry/extensions/podsecuritypolicy/storage
```
I'm surprised this package has to change - are you sure?

krmayankk replied:
there are some tests which verify storage of PodSecurityPolicy at REST layer here https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/extensions/podsecuritypolicy/storage/storage_test.go this will change if we need to validate RunAsGroup field as well

smarterclayton commented:
This is lgtm.  @pmorie / @pweil- any final comments

pweil- commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,191 @@
+  in the Docker image will be used.
+- If no `RunAsGroup` is provided in the PodSecurityContext and SecurityContext, and none in the image,
+  the container will run with primary Group as root(0).
+
```
might note here that non-numeric groups will cause an error like the behavior for `runAsUser` in the `MustRunAsNonRoot` case.

krmayankk replied:
@pweil- the api fields only take numeric values here.

pweil- replied:
@krmayankk `MustRunAsNonRoot` (for `runAsUser`) includes a kubelet check that parses the value from the dockerfile which main contain non-numeric values.  If it is a non-numeric value we error.  I think you want the same thing here.  Please correct me if I'm wrong.

krmayankk replied:
@pweil-  You are right about the check. Some additional context.  In SecurityContext RunAsNonRoot is a bool which if set to true, the docker image user is verified here https://github.com/kubernetes/kubernetes/blob/b32639f9e7d47d0f69bb41b5535803ae4c9d835a/pkg/kubelet/kuberuntime/security_context.go#L81
 It seems we will have to add an additional check to verify both user and group in that file. Also the non numeric verification you mention is here 
https://github.com/kubernetes/kubernetes/blob/9737efccf0e026e2e2437aa31c4365577d628bd5/pkg/kubelet/kuberuntime/kuberuntime_container.go#L186
 The check you mention is against SecurityContext.RunAsNonRoot and not against PSP.MustRunAsNonRoot policy but i am sure it somehow plumbs through. I will add an additional bullet to what you mention.
@pmorie @pweil-  do you both agree that RunAsNonRoot boolean in SecurityContext should apply to both uid and gid and not just user ?

pweil- replied:
> The check you mention is against SecurityContext.RunAsNonRoot and not against PSP.MustRunAsNonRoot policy but i am sure it somehow plumbs through

the strategy sets the boolean to true during admission

> @pmorie @pweil- do you both agree that RunAsNonRoot boolean in SecurityContext should apply to both uid and gid and not just user ?

hrm, that's sticky.  They are two different strategies so they could theoretically be set to different values, a single boolean won't be able to express that condition to allow the kubelet to do the check correctly

krmayankk replied:
In that case we would need another boolean `RunAsNonRootGroup` in addition to `RunAsNonRoot` which will continue to apply to uid. I will add that to the proposal

krmayankk replied:
@pmorie can you raise your concerns about not adding RunAsNonRootGroup here. @pweil- @smarterclayton i have updated the doc to also say we need `RunAsNonRootGroup` in addition to `RunAsNonRoot` similar to PSP supporting two different strategies  for Group and User. RunAsNonRoot in SecurityContext and PodSecurityContext is specific to uid and hence we need to add one one more bool RunAsNonRootGroup for gid. PTAL and comment here

php-coder replied:
> @krmayankk MustRunAsNonRoot (for runAsUser) includes a kubelet check that parses the value from the dockerfile which main contain non-numeric values. If it is a non-numeric value we error.

@krmayankk @pweil- I don't see where this lead to an error. Could you look at the code and prove that I'm wrong?

I see that:
1) in the `generateContainerConfig` we're [getting uid or username from a Dockerfile](https://github.com/kubernetes/kubernetes/blob/9737efccf0e026e2e2437aa31c4365577d628bd5/pkg/kubelet/kuberuntime/kuberuntime_container.go#L181)
2) we're [invoking `verifyRunAsNonRoot` only when uid is not nil](https://github.com/kubernetes/kubernetes/blob/9737efccf0e026e2e2437aa31c4365577d628bd5/pkg/kubelet/kuberuntime/kuberuntime_container.go#L185-L192). Otherwise, `username` is set to a non-numeric user name and we only print a warning
3) later, we're executing `generateLinuxContainerConfig` that is invoking `determineEffectiveSecurityContext`. [The latter is only using `uid` or `username`](https://github.com/kubernetes/kubernetes/blob/9737efccf0e026e2e2437aa31c4365577d628bd5/pkg/kubelet/kuberuntime/security_context.go#L42-L48) and doesn't do any check whether username is allowed or it's not.

pweil- replied:
Yes, it looks like when non-numeric support was added [here, I think](https://github.com/kubernetes/kubernetes/commit/99ee3f4b762d5a9f5d6736b62ffeba383b3d7beb#diff-08c8f9ff3cc05ba7009dc6a00cba8fa7R81) it was added as a warning.  I feel like this may be a bug though.

php-coder replied:
@Random-Liu could you comment on the aforementioned commit? Was it made by mistake or by purpose?

php-coder replied:
@Random-Liu ping ^^^^

pweil- commented on contributors/design-proposals/runas-groupid.md:
```diff
@@ -0,0 +1,191 @@
+
+- `FSGroup` and `SupplementalGroups` will continue to have their old meanings and would be untouched.  
+- The `RunAsGroup` In the SecurityContext will override the `RunAsGroup` in the PodSecurityContext.
+- If no `RunAsGroup` is provided in the PodSecurityContext and SecurityContext, the Group provided 
```
I think this is desired but might be a change in behavior that we need to call out for the feature.  Right now the docker shim uses the value of `runAsUser` in `Config.User` and doesn't try to parse anything out of the image so it would ignore anything in the dockerfile.  Similar to this example when I set the user during the run command:

```
[pweil@localhost test]$ cat Dockerfile
FROM busybox
USER 0:1000
[pweil@localhost test]$ docker run pweil/group-test id
uid=0(root) gid=1000
[pweil@localhost test]$ docker run -u 1000 pweil/group-test id
uid=1000 gid=0(root)
```

The behavior above would mean the second run would have shown `1000:1000`.

pmorie replied:
@pweil- So... we're effectively giving you group 0 whether you like it or not if you set runAsUser right now, right?

pweil- replied:
right

pweil- replied:
I guess just for completeness here, it's not really "we", the runtime behavior does it so there is argument there to say that you'll get whatever the runtime gives you unless we want to provide a consistent guarantee across implementations.

Also, it works the same way on the run command for the uid if only a group is set.  You're going to get root even if the Dockerfile has a UID:

```
[pweil@localhost test]$ cat Dockerfile 
FROM busybox
USER 1:1000
[pweil@localhost test]$ docker run -u :9999 pweil/group-test id
uid=0(root) gid=9999
```

pweil- replied:
@krmayankk I think my position here would be that we do not make a guarantee to honor the GID in the dockerfile if `runAsGroup` is unset.  We guarantee to set the values you've given us and the runtime dictates the rest.  That means:

1.  `runAsUser` set to 9999, `runAsGroup` set to 9999 -> `Config.User` set to `9999:9999`
2.  `runAsUser` set to 9999, `runAsGroup` unset -> `Config.User` set to `9999` -> docker runs you with `9999:0`
3.  `runAsUser` unset, `runAsGroup` set to 9999 -> `Config.User` set to `:9999` -> docker runs you with `0:9999`

it's probably what people expect since that is current behavior from the `-u` flag and how things run now.

krmayankk replied:
@pweil- i understand the current docker behavior but is it the right behavior ?  Its a little weird to say i will only honour the Dockerfile user when -u is NOT specified , but if specified i will completely ignore the Dockerfile and take the user or group from -u flag and make the non specified one as root(0). Why not take the non specified user or group from the Dockerfile ?

pweil- replied:
while I think that is ideal it is also a change to behavior which has backwards compatibility impacts at this point.  I don't think it's a required change to get this new api field so maybe it can be discussed more without holding up any of the progress you want to make immediately.

pmorie commented:
This is provisionally LGTM modulo the concerns about delegating to runtimwhen runasagroup is unspecified

krmayankk commented:
Sounds good to me . I will update the doc to keep the existing backward
compatible behavior .

On Wed, Sep 13, 2017 at 6:59 AM Paul Weil <notifications@github.com> wrote:

> *@pweil-* commented on this pull request.
> ------------------------------
>
> In contributors/design-proposals/runas-groupid.md
> <https://github.com/kubernetes/community/pull/756#discussion_r138627482>:
>
> > +     // container must run as a particular gid.
> +     RunAsGroupStrategyMustRunAs RunAsGroupStrategy = "MustRunAs"
> +     // container must run as a non-root gid
> +     RunAsGroupStrategyMustRunAsNonRoot RunAsGroupStrategy = "MustRunAsNonRoot"
> +     // container may make requests for any gid.
> +     RunAsGroupStrategyRunAsAny RunAsGroupStrategy = "RunAsAny"
> + )
> +```
> +
> +## Behavior
> +
> +Following points should be noted:
> +
> +- `FSGroup` and `SupplementalGroups` will continue to have their old meanings and would be untouched.
> +- The `RunAsGroup` In the SecurityContext will override the `RunAsGroup` in the PodSecurityContext.
> +- If no `RunAsGroup` is provided in the PodSecurityContext and SecurityContext, the Group provided
>
> while I think that is ideal it is also a change to behavior which has
> backwards compatibility impacts at this point. I don't think it's a
> required change to get this new api field so maybe it can be discussed more
> without holding up any of the progress you want to make immediately.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/756#discussion_r138627482>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AP4-NNbUO3auNr-VVx2b0vq1X24Ruh-iks5sh9-wgaJpZM4OB0P7>
> .
>
-- 
-Mayank

krmayankk commented:
@pweil- @pmorie addressed comments

krmayankk commented:
@pmorie @pweil- please look at the behavior section again

krmayankk commented:
This was in the outdated section so hidden. Repasting this here, so that people can continue the discussion here.
@pmorie can you raise your concerns about not adding RunAsNonRootGroup here. @pweil- @smarterclayton i have updated the doc to also say we need RunAsNonRootGroup in addition to RunAsNonRoot similar to PSP supporting two different strategies for Group and User. RunAsNonRoot in SecurityContext and PodSecurityContext is specific to uid and hence we need to add one one more bool RunAsNonRootGroup for gid. PTAL and comment here

smarterclayton commented:
I think given that people may have NFS or other shared storage that might
be impacted by root group 0, that RunAsNonRootGroup makes sense for its
clear parallel to user in the API as well us not having a good reason for
it to be inconsistent.

On Fri, Sep 15, 2017 at 5:42 PM, krmayankk <notifications@github.com> wrote:

> This was in the outdated section so hidden. Repasting this here, so that
> people can continue the discussion here.
> @pmorie <https://github.com/pmorie> can you raise your concerns about not
> adding RunAsNonRootGroup here. @pweil- <https://github.com/pweil->
> @smarterclayton <https://github.com/smarterclayton> i have updated the
> doc to also say we need RunAsNonRootGroup in addition to RunAsNonRoot
> similar to PSP supporting two different strategies for Group and User.
> RunAsNonRoot in SecurityContext and PodSecurityContext is specific to uid
> and hence we need to add one one more bool RunAsNonRootGroup for gid. PTAL
> and comment here
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/756#issuecomment-329911013>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p2mU4G8cg5M4wJpom2M9LBMVsAb-ks5siu9UgaJpZM4OB0P7>
> .
>

pmorie commented:
/lgtm

krmayankk commented:
thanks to @pmorie  for initial pointers, review and encouragement. And thanks to @pweil- @smarterclayton @vpal for reviewing and it and getting this in

bgrant0607 commented:
@krmayankk All of the design proposals moved into SIG-specific subdirectories last week. The security context proposal was moved to auth, so I'd appreciate it if this were moved into that subdirectory, also.

krmayankk commented:
Sounds good will move it today

On Mon, Sep 18, 2017 at 9:26 PM Brian Grant <notifications@github.com>
wrote:

> @krmayankk <https://github.com/krmayankk> All of the design proposals
> moved into SIG-specific subdirectories last week. The security context
> proposal was moved to auth, so I'd appreciate it if this were moved into
> that subdirectory, also.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/756#issuecomment-330425732>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AP4-NCv-VmOfWuYuMrhRk6VW1zt8YH-Fks5sj0J3gaJpZM4OB0P7>
> .
>
-- 
-Mayank
